File: frontend/src/app/interfaces/movie.ts
import { Showtime } from "@/app/interfaces/showtime";

export interface Movie {
  id: string;
  title: string;
  description: string;
  showtimes: Showtime[];
  releaseDate: string;
  poster: string;
  trailer: string;
  genres: Genre[];
  director: string;
  producer: string;
  cast: Actor[];
  synopsis: string;
  mpaa: string;
  reviews: Review[];
  upcoming: boolean;
  showing: boolean;
}

interface Genre {
  id: string;
  genre: string;
}

interface Actor {
  id: string;
  actor: string;
}

interface Review {
  id: string;
  review: string;
}

File: backend/src/main/java/com/ASCP/FlickFlare/model/Movie.java
package com.ASCP.FlickFlare.model;

import java.util.List;
import java.util.Set;

import jakarta.persistence.*;
import lombok.Data;

@Entity
@Data
@Table(name = "movie")
public class Movie {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(unique = true, nullable = false)
    private long id;
    @Column
    private String title;
    @Column
    private String description;
    @OneToMany(mappedBy = "movie", cascade = CascadeType.PERSIST, orphanRemoval = true, fetch = FetchType.EAGER)
    private List<Showtime> showtimes;
    @Column
    private long releaseDate;
    @Column
    private boolean upcoming;
    @Column
    private boolean showing;
    @Column
    private String poster;
    @Column
    private String trailer;
    @ManyToMany(cascade = CascadeType.MERGE, fetch = FetchType.EAGER)
    @JoinTable(
            name = "movie_genre",
            joinColumns = @JoinColumn(name = "movie_id"),
            inverseJoinColumns = @JoinColumn(name = "genre_id")
    )
    private Set<Genre> genres;
    @Column
    private String director;
    @Column
    private String producer;
    @ManyToMany(cascade = CascadeType.MERGE, fetch = FetchType.EAGER)
    @JoinTable(
            name = "movie_cast",
            joinColumns = @JoinColumn(name = "movie_id"),
            inverseJoinColumns = @JoinColumn(name = "actor_id")
    )
    private Set<Actor> cast;
    @Column
    private String synopsis;
    @Column
    private String mpaa;
    @OneToMany(mappedBy = "movie", cascade = CascadeType.PERSIST, orphanRemoval = true, fetch = FetchType.EAGER)
    private List<MovieReview> reviews;
}
